src/libaktualizr/primary/aktualizr.cc

```diff
@@ -45,10 +45,94 @@ void Aktualizr::systemSetup() {
 
 void Aktualizr::Initialize() { uptane_client_->initialize(); }
 
+void Aktualizr::FinalizeInstall() {
+  uptane_client_->putManifest();
+
+  boost::filesystem::path reboot_flag = "/tmp/aktualizr_reboot_flag";
+  if (boost::filesystem::exists(reboot_flag)) {
+    boost::filesystem::remove(reboot_flag);
+    if (getppid() == 1) {  // if parent process id is 1, aktualizr runs under systemd
+      exit(0);             // aktualizr service runs with 'Restart=always' systemd option.
+      // Systemd will start aktualizr automatically if it exit.
+    } else {  // Aktualizr runs from terminal
+      LOG_INFO << "Aktualizr has been updated and requires restarting to run the new version.";
+    }
+  }
+}
+
+Aktualizr::CycleEventHandler::CycleEventHandler(Aktualizr &akt) : aktualizr(akt), fut(finished.get_future()) {}
+
+void Aktualizr::CycleEventHandler::breakLoop() {
+  std::lock_guard<std::mutex> l(m);
+  if (!running) {
+    return;
+  }
+  running = false;
```

Comment:
The goal is to avoid `finished` to be set twice. If two `breakLoop()` are called from different threads, they could both pass through `if (!running)`.

Reply:
Good call, I hadn't realized that. Carry on, then!